1-Draft/RFC00XX-PredIctive-IntelliSense.md

```diff
@@ -0,0 +1,359 @@
+
+Non-goals:
+
+- Due to limit use, PowerShell Session history is not part of the historical suggestion
+- VSCode implementation. We agree this is important for a consistent experience and are working to
+  bring this feature in a future release
+
+## Specification
+
+The proposal is to add Predictive IntelliSense and Extension model to improve the user interactive experience.
+
+### Key Design Considerations Predictive IntelliSense
+
+Predictive IntelliSense displays a best match command completion from command history. The
+prediction may change if a better match is found as the user types additional information. The
+predicted command may be accepted, edited, or ignored.
+
+![Prediction - Historical](./media/pre-wlc.png)
+
+### Install and Remove Predictive IntelliSense
+
+Predictive IntelliSense is a part of the PSReadLine module. Updating or installing the current beta
+release from PSGallery will provide the Predictive IntelliSense feature.
+
+The current release is PSReadLine 2.1.0 Beta 2:
+
+```powershell
+Install-Module PSReadLine -RequiredVersion 2.1.0-beta2 -AllowPrerelease
+```
+
+It is possible to remove, delete, the PSReadLine module to disable Predictive IntelliSense. This
+creates a negative impact to customers as it would remove all features of PSReadLine. For this
+reason, Predictive IntelliSense has its own PSReadLine configuration to enable and disable
+predictions.
+
+### Enabling and Disabling Predictions
+
+By default, the feature Predictive IntelliSense is disabled. The impact of enabling this feature by
+default may cause new and existing users to become confused at the shell prompt. The feature is
+enabled and disabled with a PSReadLine command Set-PSReadLineOption that is executed at the shell
+prompt or in the user’s Profile.
+
+Predictive IntelliSense currently supports four arguments for a prediction source;
+
+* None - This option disables Predictive IntelliSense
+* History - This option uses only the PSReadLine history for predictions
+* Plugin – This option uses only registered plugin modules for predictions
+* HistoryAndPlugin – This option uses both for predictions
+
+To enable Predictive IntelliSense, enter the following command in the shell or in the users Profile:
+
+```powershell
+Set-PSReadLineOption -PredictionSource History
+```
+
+To disable Predictive IntelliSense, enter the following command in the shell or in the users Profile:
+
+```powershell
+Set-PSReadLineOption -PredictionSource None
+```
+
+### Changing the Prediction View
+
+Predictions are displayed in one of two view depending on the user preference. The default view is InlineView.
+
+* InlineView – This is the default view and displays the prediction inline with the user’s typing.
+  This view is similar to other shells Fish and ZSH.
+* ListView – ListView provides a dropdown list of predictions below the line the user is typing.
+  Users may quickly scan the list, highlight and select the desired prediction.
+
```

Comment:
Should be an option for InlineView alongside a secondary view. Inline and List complement each other. Inline can show the suggested completion with List showing multiple options at once. 

Additional Views to consider are 
* None
* Wide - the view `crtl+space` and `Format-Wide` uses
* Menu - VT Sequence Menu that looks like a real intellisense menu that you get in ISE and VS Code Editors
* TerminalIntegration - Long term, there are feature requests to provide native rendering of completion in Windows Terminal. See https://github.com/microsoft/terminal/issues/3121 and https://github.com/microsoft/terminal/issues/6632

These additional views make me think Inline should be separate On/Off with additional views as a separate configuration with None only affecting what's in its list and not Inline setting.

Reply:
@dragonwolf83 Thank you!  We are open to considering additional views as we progress. I agree that InlineView and ListView can be complimentary - currently when using ListView, as you highlight selections, they are displayed fully on the same line as your cursor, similar to InlineView. Appreciate your thoughts and feedback as we move forward.